packages/integration-sdk-runtime/src/execution/dependencyGraph.ts

```diff
@@ -242,14 +251,74 @@ export function executeStepDependencyGraph<
               metricName: `duration-step`,
               dimensions: {
                 stepId,
+                cached: hasCachePath(stepId).toString(),
+              },
+              operation: () => {
+                if (hasCachePath(stepId)) {
+                  return loadCacheForStep(step);
+                } else {
+                  return executeStep(step);
+                }
               },
-              operation: () => executeStep(step),
             }).catch(handleUnexpectedError),
           );
         }
       });
     }
 
+    async function loadCacheForStep(step: Step<TStepExecutionContext>) {
+      let status = StepResultStatus.FAILURE;
+      const stepCacheFilePath = stepStartStates[step.id].stepCachePath;
+      const entitiesPath = `${stepCacheFilePath}/entities`;
+      const relationshipsPath = `${stepCacheFilePath}/relationships`;
+
```

Comment:
I think we want to use the `jobState` interface to add entities and relationships rather than the `graphObjectsStore` to ensure some functionality is preserved (this includes use of the typeTracker, duplicateKeyTracker, and `beforeAddEntity` function, if it exists). 

```suggestion
      if (createStepGraphObjectDataUploader) {
        uploader = createStepGraphObjectDataUploader(step.id);
      }

      const { logger, jobState } = buildStepContext<
        TExecutionContext,
        TStepExecutionContext
      >({
        context: executionContext,
        duplicateKeyTracker,
        typeTracker,
        graphObjectStore,
        dataStore,
        stepId: step.id,
        beforeAddEntity,
        uploader,
      });
```

Then below: 

```diff
-           await graphObjectStore.addEntities(step.id, entities);
+           await jobState.addEntities(entities);
```

```diff
-           await graphObjectStore.addRelationships(step.id, relationships);
+           await jobState.addRelationships(relationships);
```

And finally after all have been ingested:

```diff
+ await jobState.flush();
```

Reply:
You also get access to the step-specific logger 🥳

Reply:
I'm a little bit wondering whether it's worth just injecting the `loadCacheForStep` into the `executeStep` function (rather than where it's at inside of `enqueueLeafSteps()`, so that we don't have to create the `context` in two places, and we can reuse all of the existing log messages etc. defined in `executeStep`. Probably not worth changing, but a thought exercise.

```diff
      try {
-       await step.executionHandler(context);
+       if (hasCachePath(stepId)) {
+         await loadCacheForStep(stepId, context);
+       } else {
+         await step.executionHandler(context);
+       }
        context.logger.stepSuccess(step);
```

Reply:
Great suggestion here. Really make the logging beautiful instead of 2nd class.